File: apps/backend-java/src/main/java/com/enterpriseagentos/backendjava/domain/features/management/OrgSubscription.java
package com.enterpriseagentos.backendjava.domain.features.management;

import java.time.Instant;
import java.time.ZoneOffset;
import java.time.temporal.ChronoUnit;
import java.util.UUID;
import com.enterpriseagentos.backendjava.domain.common.valueobjects.BillingPeriod;
import com.enterpriseagentos.backendjava.domain.common.valueobjects.SubscriptionPlan;

public class OrgSubscription {
    public UUID id = UUID.randomUUID();
    public String organizationId;
    public SubscriptionPlan plan = SubscriptionPlan.Free;
    public String stripeCustomerId;
    public String stripeSubscriptionId;
    public String stripeOverageItemId;
    public int concurrentAgentLimit = 1;
    public long creditBudgetPerMonth = 500_000L;
    public long creditsUsedThisMonth;
    public BillingPeriod period;
    public boolean isActive = true;
    public boolean overageEnabled;

    public static OrgSubscription createDefaultFree(String orgId) {
        PlanLimit limits = PlanLimits.ORG_FREE;
        Instant start = Instant.now().atOffset(ZoneOffset.UTC)
            .withDayOfMonth(1).truncatedTo(ChronoUnit.DAYS).toInstant();
        OrgSubscription subscription = new OrgSubscription();
        subscription.organizationId = orgId;
        subscription.plan = limits.plan();
        subscription.concurrentAgentLimit = limits.concurrentAgents();
        subscription.creditBudgetPerMonth = limits.creditsPerMonth();
        subscription.period = new BillingPeriod(start, start.plus(31, ChronoUnit.DAYS));
        return subscription;
    }

    public CreditBudgetResult checkBudget() {
        long remaining = creditBudgetPerMonth - creditsUsedThisMonth;
        return new CreditBudgetResult(remaining, remaining < 0);
    }
}

File: apps/backend/src/Features/Management/Domain/OrgSubscription.cs
namespace OffceOs.Domain.Features.Management;

public sealed class OrgSubscription
{
    public Guid Id { get; init; } = Guid.NewGuid();
    public string OrganizationId { get; init; } = string.Empty;

    public SubscriptionPlan Plan { get; set; } = SubscriptionPlan.Free;

    public string? StripeCustomerId { get; set; }
    public string? StripeSubscriptionId { get; set; }

    /// <summary>Stripe subscription item ID for the metered overage price, set when OverageEnabled=true.</summary>
    public string? StripeOverageItemId { get; set; }

    /// <summary>1 for Free, 10 for Team, custom for Enterprise.</summary>
    public int ConcurrentAgentLimit { get; set; } = 1;

    /// <summary>500_000 for Free, 25_000_000 for Team, custom for Enterprise. Normalized credits (not raw tokens).</summary>
    public long CreditBudgetPerMonth { get; set; } = 500_000;
    public long CreditsUsedThisMonth { get; set; } = 0;
    public BillingPeriod Period { get; set; }
    public bool IsActive { get; set; } = true;
    public bool OverageEnabled { get; set; } = false;

    // ── Domain logic ─────────────────────────────────────────────────────────

    /// <summary>Creates a free-tier subscription for a new organization.</summary>
    public static OrgSubscription CreateDefaultFree(string orgId)
    {
        var limits = PlanLimits.OrgFree;
        var now = new DateTime(DateTime.UtcNow.Year, DateTime.UtcNow.Month, 1, 0, 0, 0, DateTimeKind.Utc);
        return new OrgSubscription
        {
            OrganizationId = orgId,
            Plan = limits.Plan,
            ConcurrentAgentLimit = limits.ConcurrentAgents,
            CreditBudgetPerMonth = limits.CreditsPerMonth,
            Period = new BillingPeriod(now, now.AddMonths(1)),
            IsActive = true,
        };
    }

    /// <summary>Returns remaining credits and whether the budget is exceeded.</summary>
    public CreditBudgetResult CheckBudget()
    {
        var remaining = CreditBudgetPerMonth - CreditsUsedThisMonth;
        return new CreditBudgetResult(remaining, remaining < 0);
    }
}
